// lib0/dist/dom.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./pair-ab022bc3.cjs');
require('./map-28a001c9.cjs');
var dom = require('./dom-58958c04.cjs');



exports.CDATA_SECTION_NODE = dom.CDATA_SECTION_NODE;
exports.COMMENT_NODE = dom.COMMENT_NODE;
exports.DOCUMENT_FRAGMENT_NODE = dom.DOCUMENT_FRAGMENT_NODE;
exports.DOCUMENT_NODE = dom.DOCUMENT_NODE;
exports.DOCUMENT_TYPE_NODE = dom.DOCUMENT_TYPE_NODE;
exports.ELEMENT_NODE = dom.ELEMENT_NODE;
exports.TEXT_NODE = dom.TEXT_NODE;
exports.addEventListener = dom.addEventListener;
exports.addEventListeners = dom.addEventListeners;
exports.append = dom.append;
exports.appendChild = dom.appendChild;
exports.canvas = dom.canvas;
exports.checkNodeType = dom.checkNodeType;
exports.createDocumentFragment = dom.createDocumentFragment;
exports.createElement = dom.createElement;
exports.createTextNode = dom.createTextNode;
exports.doc = dom.doc;
exports.domParser = dom.domParser;
exports.element = dom.element;
exports.emitCustomEvent = dom.emitCustomEvent;
exports.fragment = dom.fragment;
exports.getElementById = dom.getElementById;
exports.insertBefore = dom.insertBefore;
exports.isParentOf = dom.isParentOf;
exports.mapToStyleString = dom.mapToStyleString;
exports.pairToStyleString = dom.pairToStyleString;
exports.pairsToStyleString = dom.pairsToStyleString;
exports.parseElement = dom.parseElement;
exports.parseFragment = dom.parseFragment;
exports.querySelector = dom.querySelector;
exports.querySelectorAll = dom.querySelectorAll;
exports.remove = dom.remove;
exports.removeEventListener = dom.removeEventListener;
exports.removeEventListeners = dom.removeEventListeners;
exports.replaceWith = dom.replaceWith;
exports.setAttributes = dom.setAttributes;
exports.setAttributesMap = dom.setAttributesMap;
exports.text = dom.text;
