// @snyk/docker-registry-v2-client/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = exports.getTags = exports.getRepos = exports.getManifest = exports.getLayer = exports.getImageSize = exports.getImageConfig = exports.getAuthTokenForEndpoint = exports.checkSupport = void 0;
const check_support_1 = require("./check-support");
Object.defineProperty(exports, "checkSupport", { enumerable: true, get: function () { return check_support_1.checkSupport; } });
const get_auth_token_for_endpoint_1 = require("./get-auth-token-for-endpoint");
Object.defineProperty(exports, "getAuthTokenForEndpoint", { enumerable: true, get: function () { return get_auth_token_for_endpoint_1.getAuthTokenForEndpoint; } });
const get_image_config_1 = require("./get-image-config");
Object.defineProperty(exports, "getImageConfig", { enumerable: true, get: function () { return get_image_config_1.getImageConfig; } });
const get_image_size_1 = require("./get-image-size");
Object.defineProperty(exports, "getImageSize", { enumerable: true, get: function () { return get_image_size_1.getImageSize; } });
const get_layer_1 = require("./get-layer");
Object.defineProperty(exports, "getLayer", { enumerable: true, get: function () { return get_layer_1.getLayer; } });
const get_manifest_1 = require("./get-manifest");
Object.defineProperty(exports, "getManifest", { enumerable: true, get: function () { return get_manifest_1.getManifest; } });
const get_repos_1 = require("./get-repos");
Object.defineProperty(exports, "getRepos", { enumerable: true, get: function () { return get_repos_1.getRepos; } });
const get_tags_1 = require("./get-tags");
Object.defineProperty(exports, "getTags", { enumerable: true, get: function () { return get_tags_1.getTags; } });
const types = require("./types");
exports.types = types;
