// snyk/dist/lib/plugins/sast/checks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackUsage = exports.getSastSettingsForOrg = void 0;
const request_1 = require("../../request");
const api_token_1 = require("../../api-token");
const config = require("../../config");
const common_1 = require("../../snyk-test/common");
async function getSastSettingsForOrg(org) {
    const response = await request_1.makeRequest({
        method: 'GET',
        headers: {
            Authorization: `token ${api_token_1.api()}`,
        },
        qs: common_1.assembleQueryString({ org }),
        url: `${config.API}/cli-config/settings/sast`,
        gzip: true,
        json: true,
    });
    return response.body;
}
exports.getSastSettingsForOrg = getSastSettingsForOrg;
async function trackUsage(org) {
    const response = await request_1.makeRequest({
        method: 'POST',
        headers: {
            Authorization: `token ${api_token_1.api()}`,
        },
        qs: common_1.assembleQueryString({ org }),
        url: `${config.API}/track-sast-usage/cli`,
        gzip: true,
        json: true,
    });
    return response.body;
}
exports.trackUsage = trackUsage;
