// snyk-docker-plugin/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateDockerfileBaseImageNameErrorCode = exports.updateDockerfileBaseImageName = exports.DockerFileAnalysisErrorCode = exports.analyseDockerfile = exports.facts = exports.dockerFile = exports.display = exports.scan = void 0;
const display_1 = require("./display");
Object.defineProperty(exports, "display", { enumerable: true, get: function () { return display_1.display; } });
const dockerFile = require("./dockerfile");
exports.dockerFile = dockerFile;
const dockerfile_1 = require("./dockerfile");
Object.defineProperty(exports, "analyseDockerfile", { enumerable: true, get: function () { return dockerfile_1.analyseDockerfile; } });
Object.defineProperty(exports, "updateDockerfileBaseImageName", { enumerable: true, get: function () { return dockerfile_1.updateDockerfileBaseImageName; } });
const types_1 = require("./dockerfile/types");
Object.defineProperty(exports, "DockerFileAnalysisErrorCode", { enumerable: true, get: function () { return types_1.DockerFileAnalysisErrorCode; } });
Object.defineProperty(exports, "UpdateDockerfileBaseImageNameErrorCode", { enumerable: true, get: function () { return types_1.UpdateDockerfileBaseImageNameErrorCode; } });
const facts = require("./facts");
exports.facts = facts;
const scan_1 = require("./scan");
Object.defineProperty(exports, "scan", { enumerable: true, get: function () { return scan_1.scan; } });
