// @yarnpkg/fslib/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xfs = exports.extendFs = exports.patchFs = exports.ZipOpenFS = exports.ZipFS = exports.VirtualFS = exports.ProxiedFS = exports.PosixFS = exports.NodeFS = exports.NoFS = exports.LazyFS = exports.JailFS = exports.CwdFS = exports.FakeFS = exports.AliasFS = exports.toFilename = exports.ppath = exports.npath = exports.Filename = exports.PortablePath = exports.DEFAULT_COMPRESSION_LEVEL = exports.normalizeLineEndings = exports.statUtils = exports.opendir = exports.LinkStrategy = void 0;
const tslib_1 = require("tslib");
const statUtils = tslib_1.__importStar(require("./statUtils"));
exports.statUtils = statUtils;
var copyPromise_1 = require("./algorithms/copyPromise");
Object.defineProperty(exports, "LinkStrategy", { enumerable: true, get: function () { return copyPromise_1.LinkStrategy; } });
var opendir_1 = require("./algorithms/opendir");
Object.defineProperty(exports, "opendir", { enumerable: true, get: function () { return opendir_1.opendir; } });
var FakeFS_1 = require("./FakeFS");
Object.defineProperty(exports, "normalizeLineEndings", { enumerable: true, get: function () { return FakeFS_1.normalizeLineEndings; } });
var ZipFS_1 = require("./ZipFS");
Object.defineProperty(exports, "DEFAULT_COMPRESSION_LEVEL", { enumerable: true, get: function () { return ZipFS_1.DEFAULT_COMPRESSION_LEVEL; } });
var path_1 = require("./path");
Object.defineProperty(exports, "PortablePath", { enumerable: true, get: function () { return path_1.PortablePath; } });
Object.defineProperty(exports, "Filename", { enumerable: true, get: function () { return path_1.Filename; } });
var path_2 = require("./path");
Object.defineProperty(exports, "npath", { enumerable: true, get: function () { return path_2.npath; } });
Object.defineProperty(exports, "ppath", { enumerable: true, get: function () { return path_2.ppath; } });
Object.defineProperty(exports, "toFilename", { enumerable: true, get: function () { return path_2.toFilename; } });
var AliasFS_1 = require("./AliasFS");
Object.defineProperty(exports, "AliasFS", { enumerable: true, get: function () { return AliasFS_1.AliasFS; } });
var FakeFS_2 = require("./FakeFS");
Object.defineProperty(exports, "FakeFS", { enumerable: true, get: function () { return FakeFS_2.FakeFS; } });
var CwdFS_1 = require("./CwdFS");
Object.defineProperty(exports, "CwdFS", { enumerable: true, get: function () { return CwdFS_1.CwdFS; } });
var JailFS_1 = require("./JailFS");
Object.defineProperty(exports, "JailFS", { enumerable: true, get: function () { return JailFS_1.JailFS; } });
var LazyFS_1 = require("./LazyFS");
Object.defineProperty(exports, "LazyFS", { enumerable: true, get: function () { return LazyFS_1.LazyFS; } });
var NoFS_1 = require("./NoFS");
Object.defineProperty(exports, "NoFS", { enumerable: true, get: function () { return NoFS_1.NoFS; } });
var NodeFS_1 = require("./NodeFS");
Object.defineProperty(exports, "NodeFS", { enumerable: true, get: function () { return NodeFS_1.NodeFS; } });
var PosixFS_1 = require("./PosixFS");
Object.defineProperty(exports, "PosixFS", { enumerable: true, get: function () { return PosixFS_1.PosixFS; } });
var ProxiedFS_1 = require("./ProxiedFS");
Object.defineProperty(exports, "ProxiedFS", { enumerable: true, get: function () { return ProxiedFS_1.ProxiedFS; } });
var VirtualFS_1 = require("./VirtualFS");
Object.defineProperty(exports, "VirtualFS", { enumerable: true, get: function () { return VirtualFS_1.VirtualFS; } });
var ZipFS_2 = require("./ZipFS");
Object.defineProperty(exports, "ZipFS", { enumerable: true, get: function () { return ZipFS_2.ZipFS; } });
var ZipOpenFS_1 = require("./ZipOpenFS");
Object.defineProperty(exports, "ZipOpenFS", { enumerable: true, get: function () { return ZipOpenFS_1.ZipOpenFS; } });
var patchFs_1 = require("./patchFs");
Object.defineProperty(exports, "patchFs", { enumerable: true, get: function () { return patchFs_1.patchFs; } });
Object.defineProperty(exports, "extendFs", { enumerable: true, get: function () { return patchFs_1.extendFs; } });
var xfs_1 = require("./xfs");
Object.defineProperty(exports, "xfs", { enumerable: true, get: function () { return xfs_1.xfs; } });
